backend/src/worker.rs

```diff
@@ -1665,6 +1662,274 @@ def main():
+                const buf = new Uint8Array([0]);
+                const sock = await Deno.connect({ port });
+                await sock.write(new Uint8Array([index]));
+                if (1 != await sock.read(buf)) throw "read";
+                return buf[0];
+            }
+            "#;
+
+            let last = r#"
+def main(last, port):
+    with __import__("socket").create_connection((None, port)) as sock:
+        sock.send(b'\xff')
+        return last + [sock.recv(1)[0]]
+            "#;
+
+            serde_json::from_value(serde_json::json!({
+                "modules": [{
+                    "value": {
+                        "type": "forloopflow",
+                        "iterator": { "type": "javascript", "expr": "result.items" },
+                        "skip_failures": false,
+                        "value": {
+                            "modules": [{
+                                "input_transform": {
+                                    "index": { "type": "javascript", "expr": "previous_result.iter.index" },
+                                    "port": { "type": "javascript", "expr": "flow_input.port" },
+                                },
+                                "value": {
+                                    "type": "rawscript",
+                                    "language": "deno",
+                                    "content": inner,
+                                },
+                            }],
+                        }
+                    }
+                }, {
+                    "input_transform": {
+                        "last": { "type": "javascript", "expr": "previous_result" },
+                        "port": { "type": "javascript", "expr": "flow_input.port" },
+                    },
+                    "value": {
+                        "type": "rawscript",
+                        "language": "python3",
+                        "content": last,
+                    },
+                }],
+                "retry": { "constant": { "attempts": 2, "seconds": 0 } },
+            })).unwrap()
+        }
+
+        #[sqlx::test(fixtures("base"))]
+        async fn test_pass(db: DB) {
+            initialize_tracing().await;
+
+            /* fails twice in the loop, then once on the last step
+             * retry attempts is measured per-step, so it _retries_ at most two times on each step,
+             * which means it may run the step three times in total */
+
+            let (attempts, responses) = [
+                /* pass fail */
+                (0, Some(99)),
+                (1, None),
+                /* pass pass fail */
+                (0, Some(99)),
+                (1, Some(99)),
+                (2, None),
+                /* pass pass pass */
+                (0, Some(3)),
+                (1, Some(5)),
+                (2, Some(7)),
+                /* fail the last step once */
+                (0xff, None),
+                (0xff, Some(9)),
```

Comment:
i've re-read the tests a few times and I couldn't make much sense of the meaning of 99, and 3,5,7 and attempts tests

Reply:
The particular numbers are arbitrary, but the values from the first two tries should be rejected so having them be 99 should help us know, if we see it, that it shouldn't be there and where it came from.

The `attempts` assert is to verify the steps are running/re-running in the right order.